pkg/meta/utils.go

```diff
@@ -494,116 +447,95 @@ func (m *baseMeta) GetTreeSummary(ctx Context, root *TreeSummary, depth, topN ui
+			tree.Files = summary.Files
+			tree.Size = summary.Size + uint64(align4K(0))
 		}
-		if err := eg.Wait(); err != nil {
-			return errno(err)
+		return err
+	}
+
+	var entries []*Entry
+	if err := m.en.doReaddir(ctx, tree.Inode, 1, &entries, -1); err != 0 {
+		return err
+	}
+	var wg sync.WaitGroup
+	tree.Children = make([]*TreeSummary, len(entries))
+	errCh := make(chan syscall.Errno, 1)
+	var err syscall.Errno
+	for i, e := range entries {
+		child := &TreeSummary{
+			Inode: e.Inode,
+			Path:  path.Join(tree.Path, string(e.Name)),
+			Type:  e.Attr.Typ,
 		}
-		var newTrees []*TreeSummary
-		for i, entries := range entriesList {
-			tree := trees[i]
-			if tree.Type != TypeDirectory {
-				continue
-			}
-			for _, e := range entries {
-				if e.Attr.Typ == TypeDirectory {
-					tree.visitRoot(func(t *TreeSummary) {
-						t.Size += uint64(align4K(0))
-						t.Dirs++
-					})
-				} else {
-					tree.visitRoot(func(t *TreeSummary) {
-						t.Size += uint64(align4K(e.Attr.Length))
-						t.Files++
-					})
-				}
-				child := &TreeSummary{
-					Inode:  e.Inode,
-					Path:   path.Join(tree.Path, string(e.Name)),
-					Type:   e.Attr.Typ,
-					parent: tree,
-				}
-				if e.Attr.Typ != TypeDirectory {
-					child.Size = uint64(align4K(e.Attr.Length))
-					child.Files = 1
-				}
-				tree.Children = append(tree.Children, child)
-			}
-			newTrees = append(newTrees, tree.Children...)
+		tree.Children[i] = child
+		if e.Attr.Typ != TypeDirectory {
+			child.Files++
+			child.Size += uint64(align4K(e.Attr.Length))
+			continue
 		}
-		// the size of children is unknown at this time, so we sort and omit them before returning
-		defer func(ts []*TreeSummary) {
-			for _, tree := range ts {
-				if tree.Type != TypeDirectory {
-					continue
-				}
-				sort.Slice(tree.Children, func(i, j int) bool {
-					return tree.Children[i].Size > tree.Children[j].Size
-				})
-				if len(tree.Children) > int(topN) {
-					omitChild := &TreeSummary{
-						Path: path.Join(tree.Path, "..."),
-						Type: TypeDirectory,
-					}
-					for _, child := range tree.Children[topN:] {
-						omitChild.Size += child.Size
-						omitChild.Files += child.Files
-						omitChild.Dirs += child.Dirs
+		select {
+		case err = <-errCh:
+			// TODO: cancel context
+			return err
+		case concurrent <- struct{}{}:
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				err := m.getTreeSummary(ctx, child, depth-1, topN, strict, concurrent)
+				<-concurrent
+				if err != 0 && err != syscall.ENOENT {
+					select {
+					case errCh <- err:
+					default:
 					}
-					tree.Children = append(tree.Children[:topN], omitChild)
 				}
+			}()
+		default:
+			if err = m.getTreeSummary(ctx, child, depth-1, topN, strict, concurrent); err != 0 && err != syscall.ENOENT {
+				return err
 			}
-		}(trees)
-		trees = newTrees
+		}
+	}
+	wg.Wait()
+	select {
+	case err = <-errCh:
+		return err
+	default:
+	}
+
+	// pick top N
+	for _, c := range tree.Children {
```

Comment:
In this implementation, the progress bar of summary does not work, as the root tree will be updated at the last.